lib/view_model.rb

```diff
@@ -234,6 +240,11 @@ def accepts_schema_version?(schema_version)
       schema_version == self.schema_version
     end
 
+    def schema_hash(schema_versions)
+      version_string = schema_versions.to_a.sort.join(',')
```

Comment:
@chrisandreae I'm trying to figure out how to calculate the hash in JS, but I'm not sure how Ruby's `sort` method deals with the array of tuples returned by `schema_versions`. I think I figured out how it works based on the documentation here: https://docs.ruby-lang.org/en/2.0.0/Array.html

> Each object in each array is compared (using the <=> operator).
>
> Arrays are compared in an “element-wise” manner; the first two elements that are not equal will determine the return value for the whole comparison.

I assume `<=>` uses standard comparisons for strings and numbers, so I can probably write a custom comparator that handles comparing tuples in this way.

Is this correct?

Alternatively, do you think it would make sense to first join the pairs before sorting? i.e.:

```
# Not sure if this is valid Ruby
version_string = schema_versions.to_a.map({ |pair| pair.join(',') }).sort.join(',')
```

That way the JS implementation could do the exact same thing (i.e. `Object.entries(schemaVersions).map(pair => pair.join(',')).sort().join(',')`).

Reply:
Now that I think about it, maybe these two sorting methods will always produce the same result? They seem more or less equivalent, but I'm not sure how to prove whether or not they are exactly equivalent.

Reply:
```
irb(main):007:0> { 'd' => 1, 'a' => 3, 'b' => 4 }.to_a
=> [["d", 1], ["a", 3], ["b", 4]]
irb(main):008:0> _.sort
=> [["a", 3], ["b", 4], ["d", 1]]
irb(main):009:0> _.join(',')
=> "a,3,b,4,d,1"
```

The Array [comparator](https://docs.ruby-lang.org/en/2.0.0/Array.html#method-i-3C-3D-3E) sorts element-wise: comparing the first element, and if they're equivalent then the second element and so on.

But yes, as you update, the source of the tuples is a map, and so the first element is structurally guaranteed to be unique: so the sort order is necessarily defined by purely by the view name.

Reply:
I don't seem to be getting the same hashes as what's in `migration_version_hash.json`. For example, if I put the `Brand` version 3 info in `irb`:

```
irb(main):040:1* {
irb(main):041:1*   "Brand": 3,
irb(main):042:1*   "BrandAuthProvider": 1,
irb(main):043:1*   "BrandCanonicalLanguage": 1,
irb(main):044:1*   "BrandDomain": 2,
irb(main):045:1*   "BrandExternalSite": 2,
irb(main):046:1*   "BrandLogo": 1,
irb(main):047:1*   "Curriculum": 1,
irb(main):048:1*   "Organization": 5,
irb(main):049:1*   "OrganizationLicensePartner": 1,
irb(main):050:1*   "OrganizationSiteCustomization": 1,
irb(main):051:1*   "Text": 1,
irb(main):052:1*   "TextTranslation": 1
irb(main):053:0> }.to_a.sort.join(',')
=> "Brand,3,BrandAuthProvider,1,BrandCanonicalLanguage,1,BrandDomain,2,BrandExternalSite,2,BrandLogo,1,Curriculum,1,Orga
nization,5,OrganizationLicensePartner,1,OrganizationSiteCustomization,1,Text,1,TextTranslation,1"
irb(main):054:0> Digest::MD5.digest(_)
=> "k3@M/\x89l\xCA#\xF2e\xBBO-O\xF9"
irb(main):055:0> Base64.urlsafe_encode64(_)
=> "azNATS-JbMoj8mW7Ty1P-Q=="
irb(main):056:0>
```

I get `azNATS-JbMoj8mW7Ty1P-Q==` as the version hash, but that data is under the key `Uzs_1YH6Cob8W-Snv0kcfA` in `migration_version_cache.json`.

Am I doing something wrong?

Reply:
> `Digest::MD5.digest(_)`

is the problem. We previously used md5 to calculate the digest, but Gordon pointed out that browsers don't include a md5 algorithm by default, so that would unnecessarily increase the bundle size. So the hash is now calculated by using the first 16 bytes of the SHA256, which the browser natively supports.

The hash is calculated from the versions map by `ViewModel.schema_hash` in the `iknow_viewmodels_gem` here:
https://github.com/iknow/iknow_view_models/blob/master/lib/view_model.rb#L249

Reply:
Oh I see. I just followed the link here from the backend PR, so I didn't realize there were further changes.

Interesting, I didn't know the browser supports SHA256 natively.

Reply:
Here's where Gordon linked me to when discussing it: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

Reply:
I think I've gotten the SHA-256 based hash calculation working now, but the data doesn't seem to be migrated when using `versions_hash`. For example, this request does not require migrated data: `/api/brands?domain=localhost.devdomain.name&versions_hash=Uzs_1YH6Cob8W-Snv0kcfA`. However, it works when using the non-hash version params (e.g. `versions[Brand]=3&...`).

@chrisandreae Are you able to reproduce this?

Reply:
Oh, I think I found the problem: The backend checks for the existence of `versions_hash`, but then actually uses `version_hash`.